### PR TITLE
New API for forceCloseHandles

### LINKED ISSUE #5786: [ArchBoard] Review the forceCloseHandles API
This API is interesting as it uses paged responses but is closer to a long running operation based on the service implementation. The API should be reviewed to be converted to a specialized LRO and potentially having different API methods for a single handle closure (won't be LRO) and a many handle closue (may be LRO).

### PR DESCRIPTION
**Pull request by alzimmermsft**
Fixes #5786 

Splits the `forceCloseHandles` API into two APIs `forceCloseHandle` and `forceCloseAllHandles`.

### CODE REVIEW

File: sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java [hunk 1]
```diff
@@ -505,6 +506,88 @@ public DirectoryAsyncClient getSubDirectoryClient(String subDirectoryName) {
         return new PagedFlux<>(() -> retriever.apply(null), retriever);
     }
 
+    /**
+     * Closes a handle on a directory or a file at the service. This is intended to be used alongside
+     * {@link #listHandles(Integer, boolean) listHanldes}.
```
Comment (conniey) [inline sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java]: spelling

Comment (JonathanGiles) [inline sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java]: Typo - but don't bother to even include the different text.

File: sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java [hunk 2]
```diff
@@ -505,6 +506,88 @@ public DirectoryAsyncClient getSubDirectoryClient(String subDirectoryName) {
...
+     *
+     * <p><strong>Code Samples</strong></p>
+     *
+     * <p>Force close handles with handles returned by get handles in recursive.</p>
```
Comment (JonathanGiles) [inline sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java]: Does this english?

File: sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java [hunk 3]
```diff
@@ -505,6 +506,88 @@ public DirectoryAsyncClient getSubDirectoryClient(String subDirectoryName) {
...
+     *
+     * {@codesnippet com.azure.storage.file.directoryAsyncClient.forceCloseHandles}
+     *
+     * <p>For more information, see the
+     * <a href="https://docs.microsoft.com/en-us/rest/api/storageservices/force-close-handles">Azure Docs</a>.</p>
+     *
+     * @param handleId Handle ID to be closed.
+     * @return An empty response.
+     */
+    public Mono<Void> forceCloseHandle(String handleId) {
```
Comment (srnagar) [inline sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java]: Wrap this with try-catch and return error in Mono.error(). Do the same for all other public methods that return Mono/Flux/PagedFlux.

File: sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java (hunk 2 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java]: Not sure I can handle this english

File: sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java [hunk 4]
```diff
@@ -505,6 +506,88 @@ public DirectoryAsyncClient getSubDirectoryClient(String subDirectoryName) {
...
+     * {@codesnippet com.azure.storage.file.directoryAsyncClient.forceCloseHandles}
+     *
+     * <p>For more information, see the
+     * <a href="https://docs.microsoft.com/en-us/rest/api/storageservices/force-close-handles">Azure Docs</a>.</p>
+     *
+     * @param recursive Flag indicating if the operation should apply to all subdirectories and files contained in the
+     * directory.
+     * @return The number of handles closed.
+     */
+    public Mono<Integer> forceCloseAllHandles(Boolean recursive) {
```
Comment (srnagar) [inline sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java]: Can this be `boolean` instead?

File: sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java [hunk 5]
```diff
@@ -505,6 +506,88 @@ public DirectoryAsyncClient getSubDirectoryClient(String subDirectoryName) {
...
+     * @param handleId Handle ID to be closed.
+     * @return An empty response.
+     */
+    public Mono<Void> forceCloseHandle(String handleId) {
+        return withContext(context -> forceCloseHandleWithResponse(handleId, context)).flatMap(FluxUtil::toMono);
+    }
+
+    /**
+     * Closes a handle on a directory or a file at the service. This is intended to be used alongside
+     * {@link #listHandles(Integer, boolean) listHanldes}.
```
Comment (srnagar) [inline sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java]:
```suggestion
     * {@link #listHandles(Integer, boolean) listHandles}.
```

File: sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java (hunk 1 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java]: Removed

Comment (alzimmermsft) [reply sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java]: Removed

File: sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java (hunk 2 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java]: Fixed

File: sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java (hunk 4 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java]: Changed

File: sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java (hunk 5 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java]: Removed

File: sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryClient.java [hunk 6]
```diff
@@ -406,32 +406,70 @@ public DirectorySetMetadataInfo setMetadata(Map<String, String> metadata) {
...
+     * <a href="https://docs.microsoft.com/en-us/rest/api/storageservices/force-close-handles">Azure Docs</a>.</p>
+     *
+     * @param recursive Flag indicating if the operation should apply to all subdirectories and files contained in the
+     * directory.
+     * @param timeout An optional timeout applied to the operation. If a response is not returned before the timeout
+     * concludes a {@link RuntimeException} will be thrown.
+     * @param context Additional context that is passed through the Http pipeline during the service call.
+     * @return The number of handles closed.
+     */
+    public int forceCloseAllHandles(boolean recursive, Duration timeout, Context context) {
```
Comment (JonathanGiles) [inline sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryClient.java]: No `withResponse` variant?

File: sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/FileClient.java [hunk 7]
```diff
@@ -848,28 +848,66 @@ public void uploadFromFile(String uploadFilePath) {
...
+     * <a href="https://docs.microsoft.com/en-us/rest/api/storageservices/force-close-handles">Azure Docs</a>.</p>
+     *
+     * @param timeout An optional timeout applied to the operation. If a response is not returned before the timeout
+     * concludes a {@link RuntimeException} will be thrown.
+     * @param context Additional context that is passed through the Http pipeline during the service call.
+     * @return The number of handles closed.
      */
-    public PagedIterable<Integer> forceCloseHandles(String handleId, Duration timeout, Context context) {
-        return new PagedIterable<>(fileAsyncClient.forceCloseHandlesWithOptionalTimeout(handleId, timeout, context));
+    public int forceCloseAllHandles(Duration timeout, Context context) {
```
Comment (JonathanGiles) [inline sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/FileClient.java]: Same here - should there be a `withResponse`?

File: sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java [hunk 8]
```diff
@@ -506,33 +507,74 @@ public DirectoryAsyncClient getSubDirectoryClient(String subDirectoryName) {
...
+     * <p>For more information, see the
+     * <a href="https://docs.microsoft.com/en-us/rest/api/storageservices/force-close-handles">Azure Docs</a>.</p>
+     *
+     * @param recursive Flag indicating if the operation should apply to all subdirectories and files contained in the
+     * directory.
+     * @return The number of handles closed.
+     */
+    public Mono<Integer> forceCloseAllHandles(boolean recursive) {
+        return fluxContext(context -> forceCloseAllHandlesWithTimeout(recursive, null, context))
+            .reduce(0, Integer::sum);
```
Comment (anuchandy) [inline sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java]: nit; alternatively we can use `count()` operator https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Flux.html#count--

Comment (alzimmermsft) [reply sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java]: Let's continue using reduce for now as we won't need to do another mapping to downcast.

File: sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryClient.java (hunk 6 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryClient.java]: Direction for this API is to just return the number of handles closed as we are abstracting away the pagedness of the REST API.

File: sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/FileClient.java (hunk 7 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/FileClient.java]: Direction for this API is to just return the number of handles closed as we are abstracting away the pagedness of the REST API.

File: sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java (hunk 8 above)
Comment (anuchandy) [reply sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java]: Sounds good.

File: sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java (hunk 3 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java]: Done

**Pull request merged**